src/extra/components/xgm.py

```diff
@@ -0,0 +1,498 @@
+                1 and SASE 3 this can be either `1` or `3` or `0`. Passing `0`
+                is a special case that refers to the main properties of an XGM,
+                i.e. it will look at `data.intensityTD` instead of
+                `data.intensitySa1TD` with `sase=1`. This setting overrides the
+                `default_sase` argument to the
+                [constructor][extra.components.XGM].
+        """
+        pg = self._check_sase_arg(sase)
+        if pg not in self._intensity:
+            if pg == PropertyGroup.MAIN:
+                key = "data.intensityTD"
+            else:
+                key = f"data.intensitySa{pg.value}TD"
+
+            max_pulses = self.max_pulses(sase)
+            # We could use .xarray(roi=...) here, but that fails because of h5py
+            # when max_pulses is 0.
+            intensity = self.instrument_source[key].xarray()[:, :max_pulses]
+            # Assign a pulseId dimension and coordinate
+            intensity = intensity.rename(dim_0="pulseId").assign_coords(pulseId=np.arange(max_pulses))
+            # Replace the default fill value of 1 with 0
+            intensity.data[intensity.data == 1] = 0
+
+            self._intensity[pg] = intensity
+
+        return self._intensity[pg]
+
+    def npulses(self, sase=None):
+        """Return a 1D array of the number of pulses in each train.
+
+        Args:
+            sase (int): Same meaning as in
+                [XGM.intensity()][extra.components.XGM.intensity].
+        """
+        pg = self._check_sase_arg(sase)
+        if pg not in self._npulses:
+            if pg == PropertyGroup.MAIN:
+                key = "pulseEnergy.numberOfBunchesActual"
+            else:
+                key = f"pulseEnergy.numberOfSa{pg.value}BunchesActual"
+            self._npulses[pg] = self._control_source[key].xarray()
+
+        return self._npulses[pg]
+
+    def mean_pulses(self, sase=None) -> float:
+        """The mean number of pulses.
+
+        This will differ from
+        [XGM.max_pulses()][extra.components.XGM.max_pulses] if there is a
+        varying number of pulses.
+
+        Args:
+            sase (int): Same meaning as in
+                [XGM.intensity()][extra.components.XGM.intensity].
+        """
+        pg = self._check_sase_arg(sase)
+        if pg not in self._mean_pulses:
+            self._mean_pulses[pg] = self.npulses(sase).mean().item()
+
+        return self._mean_pulses[pg]
+
+    def max_pulses(self, sase=None) -> int:
+        """The maximum number of pulses.
+
+        Args:
+            sase (int): Same meaning as in
+                [XGM.intensity()][extra.components.XGM.intensity].
+        """
+        pg = self._check_sase_arg(sase)
+        if pg not in self._max_pulses:
+            self._max_pulses[pg] = int(self.npulses(sase).max().item())
+
+        return self._max_pulses[pg]
+
+    def npulses_is_constant(self, sase=None) -> bool:
```

Comment:
Yeah I think you're right, I replaced `mean_pulses()` with `nominal_pulses()` in df3e138. Not entirely sold on the naming, another option would be to have e.g. `npulses(all_trains=False)` but I'm not sure I like that API much better.

Reply:
I think I agree with you on having a separate method rather than a parameter, but I'm also not sure about the naming.

Maybe `.npulses()` should be the one giving you a single number (I'm imagining that works for >90% of experiments, but that guess might be wrong) and have a longer name like `npulses_by_train()` for getting an array?

The closest parallel I can think of with what we already have is in the multi-module detector components. They have a method `.frames_per_train()` to get a single number if possible, and a pre-computed pandas series as `.frame_counts` for the count in each train. But that's different enough that I don't think we need to copy it.

Reply:
That sounds good to me, renamed in 3a40ea3.